url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/com/serotonin/bacnet4j/type/primitive/Time.java

```diff
@@ -0,0 +1,253 @@
+ * <br>
+ * Neither an unspecified time nor a time pattern shall be used in time values<br> 
+ * that convey actual time, such as those presented by the Local_Time property<br> 
+ * of the Device object or in a TimeSynchronization-Request.<br>
+ * <br>
+ * Example: Application-tagged specific time value<br>
+ * ASN.1 = Time<br>
+ * Value = 17:35:45.17 (= 5:35:45.17 P.M.)<br>
+ * Application Tag = Time (Tag Number = 11)<br> 
+ * Encoded Tag = X'B4'<br>
+ * Encoded Data = X'11232D11'<br>
+ */
+public class Time extends Primitive {
+    private static final long serialVersionUID = -5256831366663750858L;
+
+    public static final byte TYPE_ID = 11;
+
+    private final int hour;
+    private final int minute;
+    private final int second;
+    private final int hundredth;
+
+
+    public Time(int hour, int minute, int second, int hundredth) {
+        this.hour = hour;
+        this.minute = minute;
+        this.second = second;
+        this.hundredth = hundredth;
+    }
+
+    public Time() {
+        this(new GregorianCalendar());
+    }
+
+    public Time(final GregorianCalendar now) {
+        this.hour = now.get(Calendar.HOUR_OF_DAY);
+        this.minute = now.get(Calendar.MINUTE);
+        this.second = now.get(Calendar.SECOND);
+        this.hundredth = now.get(Calendar.MILLISECOND) / 10;
+    }
+
+    public boolean isHourUnspecified() {
+        return hour == 255;
+    }
+
+    public int getHour() {
+        return hour;
+    }
+
+    public boolean isMinuteUnspecified() {
+        return minute == 255;
+    }
+
+    public int getMinute() {
+        return minute;
+    }
+
+    public boolean isSecondUnspecified() {
+        return second == 255;
+    }
+
+    public int getSecond() {
+        return second;
+    }
+
+    public boolean isHundredthUnspecified() {
+        return hundredth == 255;
+    }
+
+    public int getHundredth() {
+        return hundredth;
+    }
+
+    /**
+     * @param that
+     *            The time with which to compare this
+     * @return true if this < that.
+     */
+    public boolean before(final Time that) {
+        if (!this.isHourUnspecified() && !that.isHourUnspecified()) {
+            if (this.hour < that.hour)
+                return true;
+            if (this.hour > that.hour)
+                return false;
+        }
+
+        if (!this.isMinuteUnspecified() && !that.isMinuteUnspecified()) {
+            if (this.minute < that.minute)
+                return true;
+            if (this.minute > that.minute)
+                return false;
+        }
+
+        if (!this.isSecondUnspecified() && !that.isSecondUnspecified()) {
+            if (this.second < that.second)
+                return true;
+            if (this.second > that.second)
+                return false;
+        }
+
+        if (this.isHundredthUnspecified() || that.isHundredthUnspecified())
+            return false;
+
+        return this.hundredth < that.hundredth;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 127:151, 160:184.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.